design/ErrorHandling.md

```diff
@@ -80,90 +79,244 @@ device.addEventListener('validationerror', (event) => {
+  - `GPUBuffer.mapReadAsync`
+  - `GPUBuffer.mapWriteAsync`
+
+If there is an error, the returned Promise rejects.
+
+#### Void-Returning
+
+  - `GPUQueue.submit`
+  - `GPUQueue.signal`
+  - `GPUBuffer.setSubData`
+  - `GPUBuffer.unmap`
+  - `GPUBuffer.destroy`
+  - `GPUTexture.destroy`
+  - `GPUFence.getCompletedValue`
+
+See "Operation Status" below.
+
+## Object Status
+
+As described above, objects may be internally "valid" or "invalid".
+For testing, it is necessary to be able to inspect this status.
+Through this mechanism, allocation failure ("out-of-memory") can also be expressed.
+
+Object status can be queried with `getStatus`.
+`getStatus` returns a Promise which resolves to a `GPUObjectStatus` object that reports the status of the object.
+`GPUObjectStatus` includes a `type`, which describes the general category of error, as well as a `message` (an implementation-defined human-readable message string about the error).
+
+`getStatus` rejects if the device is lost.
+
+```webidl
+enum GPUErrorType {
+    "invalid-object", // `this` or an object argument was invalid.
+    "invalid-value",  // A non-object argument was wrong (out of range, or invalid enum).
+    "invalid-state",  // `this` or an object argument was in a wrong state (e.g. mapped).
+    // TODO: more?
+    "out-of-memory"   // An allocation failed nonfatally.
+};
+
+interface GPUObjectStatus {
+    readonly attribute boolean valid;
+
+    // These are only set if valid == false.
+    readonly attribute GPUErrorType? type;
+    readonly attribute DOMString? message;
+};
+
+typedef (GPUBindGroup
+      or GPUBindGroupLayout
+      or GPUBuffer
+      or GPUCommandBuffer
+      or GPUCommandEncoder
+      or GPUComputePipeline
+      or GPUFence
+      or GPUPipelineLayout
+      or GPUProgrammablePassEncoder
+      or GPUQueue
+      or GPURenderPipeline
+      or GPUSampler
+      or GPUShaderModule
+      or GPUSwapChain
+      or GPUTexture
+      or GPUTextureView) GPUStatusableObject;
+
+partial interface GPUDevice {
+    Promise<GPUObjectStatus> getObjectStatus(GPUStatusableObject object);
 };
 ```
 
-(Note that object creation failures always send an error to the error log, regardless of the object type or the type of failure.)
+#### Alternatives
 
-If an application uses recoverable allocation, the implementation will still generate error log entries:
-a `"recoverable-out-of-memory"` error for the object creation, and `"validation-error"`s for any subsequent uses of the invalid object.
-The application may need to understand whether such error log entries were part of a recovered allocation (e.g. to avoid sending telemetry for those errors).
-To facilitate this filtering, the handle to the invalid object (including "expando" JavaScript properties) is attached to the error log entry (see above).
+- Instead of `device.getStatus(obj)`, have `obj.getStatus()`. Con: Pollutes objects with rarely-used methods. Pro: Maybe slightly more clear to users.
```

Comment:
That is because I changed it and didn't do a good job of fixing up the references. Thanks for catching this.

Reply:
> Con: Pollutes objects with rarely-used methods.

Have we thought about having GPUStatusableObject as a base class object (that has getStatus() method) for the GPU* objects in the typedef?

Reply:
I considered it, but I'm more concerned with polluting actual objects, not polluting the IDL. This would still put one on every object.

I think I will be moving away from getStatus in the next iteration though.